core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java

```diff
@@ -67,23 +73,94 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
     }
 
     /**
+     * Creates a query builder given a query provided as a bytes array
+     */
+    public WrapperQueryBuilder(byte[] source) {
+        this(source, 0, source.length);
+    }
+
+    /**
      * Creates a query builder given a query provided as a {@link BytesReference}
      */
     public WrapperQueryBuilder(BytesReference source) {
-        this.source = source.array();
+        this.source = source.toBytes();
         this.offset = source.arrayOffset();
         this.length = source.length();
     }
 
+    public byte[] source() {
+        return this.source;
+    }
+
+    public int offset() {
+        return this.offset;
+    }
+
+    public int length() {
+        return this.length;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
         builder.field("query", source, offset, length);
+        printBoostAndQueryName(builder);
         builder.endObject();
     }
 
     @Override
-    public String getName() {
-        return NAME;
+    protected Query doToQuery(QueryParseContext parseContext) throws IOException {
+        try (XContentParser qSourceParser = XContentFactory.xContent(source).createParser(source)) {
+            final QueryParseContext context = new QueryParseContext(parseContext.index(), parseContext.indexQueryParserService());
+            context.reset(qSourceParser);
+            Query query = context.parseInnerQuery();
+            parseContext.combineNamedFilters(context);
+            return query;
+        }
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.source == null) {
+            validationException = addValidationError("query source text cannot be null", validationException);
+        }
+        if (this.offset < 0) {
+            validationException = addValidationError("offset cannot be negative.", validationException);
+        }
+        if (this.length < 0) {
+            validationException = addValidationError("length cannot be negative.", validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    protected WrapperQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        return new WrapperQueryBuilder(in.readByteArray(), in.readVInt(), in.readVInt());
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeByteArray(this.source);
+        out.writeVInt(this.offset);
+        out.writeVInt(this.length);
```

Comment:
Why not keep the constructor as is, but on serialisation only write out the part of the byte array that we are interested in (and on reading it initalise offset to zero and length to the length of the byte array we've read) like we already do in doXContent?